### PR TITLE
Introduce the `optunahub.benchmarks` module

### PR DESCRIPTION
**Pull request by y0z**
## Motivation & Description of the changes

- Introduce the `optunahub.benchmarks` module as a new feature for benchmarking.
- Update the doc files to introduce the module.

image

### CODE REVIEW

File: optunahub/benchmarks/_base_problem.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        return self.evaluate(params)
+
+    def evaluate(self, params: dict[str, Any]) -> float | Sequence[float]:
+        """Evaluate the objective function.
+
+        Args:
+            params: Dictionary of input parameters.
+
+        Returns:
+            The objective value.
```
Comment (nabenabe0928) [inline optunahub/benchmarks/_base_problem.py:35]:
```suggestion
            The objective value or a sequence of the objective values for multi-objective optimization.
```

File: optunahub/benchmarks/_base_problem.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+class BaseProblem(metaclass=ABCMeta):
+    """Base class for optimization problems."""
+
+    def __call__(self, trial: optuna.Trial) -> float | Sequence[float]:
+        """Objective function for Optuna. By default, this method calls :meth:`evaluate` with the parameters defined in :attr:`search_space`.
+
+        Args:
+            trial: Optuna trial object.
+        Returns:
+            The objective value.
```
Comment (nabenabe0928) [inline optunahub/benchmarks/_base_problem.py:20]:
```suggestion
            The objective value or a sequence of the objective values for multi-objective optimization.
```

File: optunahub/benchmarks/_base_problem.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        """
+        params = {}
+        for name, dist in self.search_space.items():
+            params[name] = trial._suggest(name, dist)
+            trial._check_distribution(name, dist)
+        return self.evaluate(params)
+
+    def evaluate(self, params: dict[str, Any]) -> float | Sequence[float]:
```
Comment (nabenabe0928) [inline optunahub/benchmarks/_base_problem.py:28]: Is there any reason why this method is not an abstract method?

File: optunahub/benchmarks/_base_problem.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+                def evaluate(self, params: dict[str, Any]) -> float:
+                    x = params["x"]
+                    y = params["y"]
+                    return x ** 2 + y
+        """
+        raise NotImplementedError
+
+    @property
+    def search_space(self) -> dict[str, optuna.distributions.BaseDistribution]:
```
Comment (nabenabe0928) [inline optunahub/benchmarks/_base_problem.py:48]: Is there any reason why this method is not an abstract method?

File: tests/test_benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        @property
+        def search_space(self) -> dict[str, optuna.distributions.BaseDistribution]:
+            return {"x": optuna.distributions.FloatDistribution(low=-1, high=1)}
+
+        @property
+        def directions(self) -> list[optuna.study.StudyDirection]:
+            return [optuna.study.StudyDirection.MINIMIZE]
+
+    problem = TestProblem()
+    study = optuna.create_study()
```
Comment (nabenabe0928) [inline tests/test_benchmarks.py:23]:
```suggestion
    study = optuna.create_study(directions=problem.directions)
```

**Review by nabenabe0928 — COMMENTED** [review]
Thank you for the PR, I left some comments!

File: optunahub/benchmarks/_base_problem.py (hunk 3 above)
Comment (y0z) [reply optunahub/benchmarks/_base_problem.py:28]: If you only want to support Optuna (e.g., implementing define-by-run objective), it is acceptable to override `__call__` and not implement `evaluate`.

File: optunahub/benchmarks/_base_problem.py (hunk 4 above)
Comment (y0z) [reply optunahub/benchmarks/_base_problem.py:48]: Ditto (cf. https://github.com/optuna/optunahub/pull/73#discussion_r1879199676)

**Review by nabenabe0928 — COMMENTED** [review]
Thank you for the update, LGTM!

**Review by nabenabe0928 — APPROVED** [review]

**Pull request merged**